README.md
# phoenix_gen_socket_client

![CI](https://github.com/alboratech/phoenix_gen_socket_client/workflows/CI/badge.svg)
[![hex.pm](https://img.shields.io/hexpm/v/phx_gen_socket_client.svg?style=flat-square)](https://hex.pm/packages/phoenix_gen_socket_client)
[![hexdocs.pm](https://img.shields.io/badge/docs-latest-green.svg?style=flat-square)](https://hexdocs.pm/phx_gen_socket_client/)

> **Fork of the [phoenix_gen_socket_client][phoenix_gen_socket_client] library.**

This library implements an Elixir client for Phoenix Channels protocol 2.x.
The client is implemented as a behaviour, which allows a lot of flexibility.
For an alternative approach, you may also want to check
[this project](https://github.com/mobileoverlord/phoenix_channel_client).

__NOTE__: From version 2.0 onwards this library only supports version 2.0 of
the channels protocol (used on Phoenix 1.3 or later). If you need to use the
version 1.0 of the protocol, you need to use
[the older version of this library](https://hex.pm/packages/phoenix_gen_socket_client/1.2.0).

[phoenix_gen_socket_client]: https://github.com/Aircloak/phoenix_gen_socket_client


## Installation

You need to add the project as a dependency to your `mix.exs`:

```elixir

def project do
  [
    deps: [
      {:phoenix_gen_socket_client, "~> 3.3.0", hex: :phx_gen_socket_client}
      # ...
    ],
    # ...
  ]
end
```

You also need to add the transport (e.g. a websocket client), and serializer
(e.g. JSON) as dependencies. Out of the box, the adapter for
[sanmiguel/websocket_client][websocket_client], and support for
[Jason](https://hex.pm/packages/jason) is provided but you still
need to include the libraries yourself:

```elixir
def project do
  [
    deps: [
      {:websocket_client, "~> 1.2"},
      {:jason, "~> 1.1"}

      # ...
    ],
    # ...
  ]
end
```

[websocket_client]: https://github.com/sanmiguel/websocket_client

## Usage

This library is designed for flexibility, so the usage is a bit more involved.
In general, you need to:

1. Implement a callback module for the `Phoenix.Channels.GenSocketClient`
   behaviour.
2. Start the socket process somewhere in your supervision tree.
3. Interact with the server from callback functions which are running in the
   socket process.

A simple demo is available [here][demo]. Here, we'll present some general ideas.
For more details, refer to the [documentation][documentation].

__Note__: In the subsequent code snippets we assume that
`Phoenix.Channels.GenSocketClient` is aliased, so we use `GenSocketClient`.

[demo]: https://github.com/alboratech/phoenix_gen_socket_client/tree/master/example
[documentation]: https://hexdocs.pm/phx_gen_socket_client/Phoenix.Channels.GenSocketClient.html

### Starting the socket process

Starting the process works similar to other behaviours, such as `GenServer`:

```elixir
GenSocketClient.start_link(
  callback_module,
  transport_module,
  arbitrary_argument,
  socket_opts, # defaults to []
  gen_server_opts # defaults to []
)
```

For `transport_module` you can pass any module which implements the
`Phoenix.Channels.GenSocketClient.Transport` behaviour. Out of the box, you have
the module `Phoenix.Channels.GenSocketClient.Transport.WebSocketClient`
available.

The code above will start another process where the `init/1` function of the
`callback_module` is invoked. This function needs to provide the initial state
and the socket url. The tuple also determines whether the connection will be
immediately established or not.

The socket url must also include the transport suffix. For example, if in the
server socket you have declared socket with `socket "/my_socket", ...`, then
the url for the websocket transport would be
`ws://server_url/my_socket/websocket`.


### Connection life-cycle

If `init/1` returns `{:connect, url, query_params, initial_state}` the socket
process will try to connect to the server. The connection is established in a
separate process which we call the transport process. This process is the
immediate child of the socket process. As a consequence, all communication takes
place concurrently to the socket process. If you handle Erlang messages in the
socket process you may need to keep track of whether you're connected or not.

The establishing of the connection is done asynchronously. The
`handle_connected/2` callback is invoked after the connection is established.
The `handle_disconnected/2` callback is invoked if establishing the connection
fails or an existing connection is lost.

If the connection is not established (or dropped), you can reconnect from
`handle_*` functions by returning `{:connect, state}` tuple. In this case the
workflow is the same as when returning the `:connect` tuple from the `init/1`
callback.

You may also decide to defer connecting to a later point in time by returning
`{:noconnect, url, query_params, state}` from the `init/1` callback. To later
establish a connection you need to send some message to the socket process,
and handle that message in `handle_info` by returning the `{:connect, state}`
tuple.

You may also return `{:connect, url, query_params, state}` from a `handle_info`
callback to connect (or reconnect) the socket to a different location or with
different params than it started with.


Though somewhat elaborate, this approach has following benefits:

1. The socket process starts immediately without waiting for the connection to
   be established.
2. The socket process can live in the supervision tree even if the connection
   is not established.
3. Implementing reconnection logic is very flexible. You can trigger a reconnect
   directly from `handle_disconnected/2`, or send yourself a delayed message.
   You can also easily accumulate outgoing messages until you reconnect, or you
   can attempt a finite number of reconnects and then give up.


### Sending messages over a socket

Once you're connected to the socket, you can issue messages. Most `handle_*`
callbacks receive a `transport` argument. You can pass this argument to
functions such as `GenSocketClient.join/3`, `GenSocketClient.leave/3`,
`GenSocketClient.push/4` to send messages over a connected socket. If you're
disconnected when calling these functions, they will return an error. Otherwise
they return a successful response.

Of course, there's no guarantees that the message has arrived to the server,
since a netsplit can happen at any point in time.


### Channel life-cycle

To join a topic, you can use `GenSocketClient.join/3`. Depending on the outcome,
`handle_joined/4` or `handle_join_error/4` callback is invoked. If the server
leaves the channel at some later point, `handle_channel_closed/4` callback will
be invoked. Just like with socket connections, this gives you a lot of
flexibility if you want to implement the rejoin logic. On socket disconnect,
all channels are closed, but `handle_channel_closed/4` is not invoked.

You can of course join multiple topics on the same socket. Unlike in Phoenix,
all messages are handled in the socket process. Besides the socket and the
transport process, there are no additional processes created. Such decisions
are deliberately left out of the `GenSocketClient` to support more flexibility.


### Communicating on a channel

To send a message on the connected channel, you can use
`GenSocketClient.push/4`. If the socket has not joined the topic, an error is
returned. On success, the message reference (aka _ref_) is returned.

If the server-side channel replies directly (using the `{:reply, ...}`), the
`handle_reply/5` callback is invoked with the matching ref. Refs are unique
only per channel-session. Two channels can use the same refs. Also, refs can be
reused after you rejoin a topic.

If the server sends an asynchronous message (i.e. not a direct reply), the
`handle_message/5` callback is invoked.


## Copyright and License

Copyright (c) 2020 Aircloak

The source code is licensed under the [MIT License](https://github.com/alboratech/phoenix_gen_socket_client/blob/master/LICENSE.md).


CHANGELOG.md
Changelog
=========

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 3.3.0

New fork maintained by Albora Tech.

### Added

- Added coveralls.
- Added dialyzer via `:dialyxir`.

### Changed

- Fixed socket client to allow passing options to serializer for decoding based
  on the opcode (`:text` or `:binary`).
- Fixed docs and format.

## 3.2.2

This library has been retired.
It is no longer maintained.
It will not receive security fixes or updates.

For more information, please read this issue:
https://github.com/Aircloak/phoenix_gen_socket_client/issues/57


## 3.2.1

- Fixed text encoding bug when upgrading Phoenix to 1.5.7 (issue #54). Thanks @vladra

## 3.2.0

- Allow passing `:extra_headers` and `:ssl_verify` options to websocket_client. Thanks @albertored

## 3.1.0

- Fix timeout typespec in GenSocketClient call function. Thanks @matt-mazzucato
- Add support for optional terminate callback

## 3.0.0

- The provided Json serializer uses Jason rather than Poison. This is a breaking change and the cause of the major version bump.
- Minor fixes to doc and tests
- Tests against more recent Elixir (1.10) and Erlang (22) versions
- Dropping support for Elixir versions prior to 1.8 and Erlang prior to 22.

## 2.1.1

- Uses credo only in `:dev` and `:test`, so it's not treated as a transitive dependency in client projects.

## 2.1.0

- Added `Phoenix.Channels.GenSocketClient.joined?/1`
- Added support for changing url and/or query parameters when reconnecting

## 2.0.0

- requires Elixir 1.5 or greater
- supports only Phoenix channels protocol 2.0 (so the library works only with Phoenix 1.3+ on sockets which are powered by 2.0 serializers)
- takes URL parameters separately (see spec for `Phoenix.Channels.GenSocketClient.init/1` and `Phoenix.Channels.GenSocketClient.TestSocket.start_link/1`)
- properly enforces optional dependencies (Poison and websocket_client)

## 1.2.0

- Serialization failures are propagated as tuples instead of exceptions

## 1.1.1

- Changed the internals to simplify creating a large number of open connections.

## 1.1.0

- Added `call/3` to support synchronous calls to the socket process.
- The behavior now mandates the `handle_call/4` callback function.


test/test_helper.exs
Phoenix.PubSub.Supervisor.start_link(
  name: TestSite.PubSub,
  adapter: Phoenix.PubSub.PG2
)

ExUnit.start()


test/support/test_site.ex
defmodule TestSite do
  @moduledoc false

  defmodule PubSub do
    @moduledoc false
    def start_link(), do: Registry.start_link(keys: :duplicate, name: __MODULE__)

    def subscribe(subscriber_key), do: Registry.register(__MODULE__, subscriber_key, nil)

    def notify(subscriber_key, message) do
      __MODULE__
      |> Registry.lookup(subscriber_key)
      |> Enum.map(fn {pid, _value} -> pid end)
      |> Enum.each(&send(&1, message))
    end
  end

  defmodule Endpoint do
    @moduledoc false
    use Phoenix.Endpoint, otp_app: :phoenix_gen_socket_client

    socket("/test_socket", TestSite.Socket,
      websocket: true,
      longpoll: false
    )

    socket("/test_socket_updated", TestSite.SocketUpdated,
      websocket: true,
      longpoll: false
    )

    @doc false
    def init(:supervisor, config) do
      {:ok,
       Keyword.merge(
         config,
         https: false,
         http: [port: 29_876],
         secret_key_base: String.duplicate("abcdefgh", 8),
         debug_errors: false,
         server: true,
         pubsub_server: TestSite.PubSub
       )}
    end
  end

  defmodule Socket do
    @moduledoc false
    use Phoenix.Socket

    # List of exposed channels
    channel("channel:*", TestSite.Channel)

    def connect(params, socket) do
      case params["shared_secret"] do
        "supersecret" -> {:ok, socket}
        _ -> :error
      end
    end

    def id(_socket), do: ""
  end

  defmodule SocketUpdated do
    @moduledoc false
    use Phoenix.Socket

    # List of exposed channels
    channel("channel:*", TestSite.Channel)

    def connect(params, socket) do
      case params["shared_secret"] do
        "supersecret_updated" -> {:ok, socket}
        _ -> :error
      end
    end

    def id(_socket), do: ""
  end

  defmodule Channel do
    @moduledoc false
    use Phoenix.Channel
    alias TestSite.PubSub

    def subscribe(), do: PubSub.subscribe(__MODULE__)

    def join(topic, join_payload, socket) do
      notify({:join, topic, join_payload, self()})
      {:ok, socket}
    end

    def handle_info({:push, event, payload}, socket) do
      push(socket, event, payload)
      {:noreply, socket}
    end

    def handle_info({:stop, reason}, socket), do: {:stop, reason, socket}
    def handle_info({:crash, reason}, _socket), do: exit(reason)

    def handle_in("sync_event", payload, socket) do
      {:reply, {:ok, payload}, socket}
    end

    def handle_in(event, payload, socket) do
      notify({:handle_in, event, payload})
      {:noreply, socket}
    end

    def terminate(reason, _socket), do: notify({:terminate, reason})

    defp notify(message), do: PubSub.notify(__MODULE__, {__MODULE__, message})
  end
end


test/socket_client_test.exs
defmodule Phoenix.Channels.GenSocketClientTest do
  use ExUnit.Case, async: false

  alias Phoenix.Channels.GenSocketClient.TestSocket
  alias TestSite.Endpoint

  setup_all do
    ExUnit.CaptureLog.capture_log(fn -> Endpoint.start_link() end)
    :ok
  end

  setup do
    TestSite.Channel.subscribe()
    :ok
  end

  test "connection success" do
    assert {:ok, socket} = start_socket()
    assert :connected == TestSocket.wait_connect_status(socket)
    assert {:ok, {"channel:1", %{}}} == TestSocket.join(socket, "channel:1")
  end

  test "no auto connect" do
    assert {:ok, socket} = start_socket(url(), query_params(), false)
    refute :connected == TestSocket.wait_connect_status(socket, 100)
    TestSocket.connect(socket)
    assert :connected == TestSocket.wait_connect_status(socket)
    assert {:ok, {"channel:1", %{}}} == TestSocket.join(socket, "channel:1")
  end

  test "connect with url and query_params" do
    # start the socket. false means do not connect for now.
    assert {:ok, socket} = start_socket(url(), query_params(), false)
    refute :connected == TestSocket.wait_connect_status(socket, 100)
    # connect by explicitly setting the url and query_params.
    TestSocket.connect(socket, url_updated(), query_params_updated())
    assert :connected == TestSocket.wait_connect_status(socket)
    assert {:ok, {"channel:1", %{}}} == TestSocket.join(socket, "channel:1")
  end

  test "client message push" do
    assert {:ok, socket} = start_socket()
    assert :connected == TestSocket.wait_connect_status(socket)
    assert {:ok, {"channel:1", %{}}} == TestSocket.join(socket, "channel:1", %{"foo" => "bar"})
    assert {:ok, _ref} = TestSocket.push(socket, "channel:1", "some_event", %{"foo" => "bar"})
    assert_receive {TestSite.Channel, {:handle_in, "some_event", %{"foo" => "bar"}}}
  end

  test "send and response" do
    conn = join_channel()

    {:ok, payload} =
      TestSocket.push_sync(conn.socket, "channel:1", "sync_event", %{"foo" => "bar"})

    assert %{"status" => "ok", "response" => %{"foo" => "bar"}} = payload
  end

  test "sending a mesasge that cannot be encoded" do
    conn = join_channel()

    assert {:error,
            {:encoding_error, %Jason.EncodeError{message: "invalid byte 0xA6 in <<166>>"}}} =
             TestSocket.push_sync(conn.socket, "channel:1", "sync_event", %{
               "foo" => _invalid_string = <<166>>
             })
  end

  test "client message receive" do
    conn = join_channel()
    send(conn.server_channel, {:push, "some_event", %{"foo" => "bar"}})

    assert {:ok, {"channel:1", "some_event", %{"foo" => "bar"}}} =
             TestSocket.await_message(conn.socket)
  end

  test "leave the channel" do
    conn = join_channel()
    assert {:ok, %{}} == TestSocket.leave(conn.socket, "channel:1")
    assert_receive {TestSite.Channel, {:terminate, {:shutdown, :left}}}
  end

  test "push references are monotonically increasing on the same channel" do
    conn = join_channel()
    {:ok, ref1} = TestSocket.push(conn.socket, "channel:1", "event1")
    {:ok, ref2} = TestSocket.push(conn.socket, "channel:1", "event2")

    assert ref1 < ref2
  end

  test "push references are monotonically increasing on multiple channels" do
    conn = join_channel()
    {:ok, _} = TestSocket.join(conn.socket, "channel:2")
    {:ok, ref1} = TestSocket.push(conn.socket, "channel:1", "event1")
    {:ok, ref2} = TestSocket.push(conn.socket, "channel:2", "event2")

    assert ref1 < ref2
  end

  test "push references are monotonically increasing after leave/rejoin" do
    conn = join_channel()
    {:ok, ref1} = TestSocket.push(conn.socket, "channel:1", "event1")
    {:ok, %{}} = TestSocket.leave(conn.socket, "channel:1")
    TestSocket.join(conn.socket, "channel:1")
    {:ok, ref2} = TestSocket.push(conn.socket, "channel:1", "event2")

    assert ref1 < ref2
  end

  test "connection error" do
    assert {:ok, socket} = start_socket("ws://127.0.0.1:29877")
    assert {:disconnected, :econnrefused} == TestSocket.wait_connect_status(socket)
  end

  test "connection refused by socket" do
    assert {:ok, socket} = start_socket(url(), shared_secret: "invalid_secret")
    assert {:disconnected, {403, "Forbidden"}} == TestSocket.wait_connect_status(socket)
  end

  test "transport process terminates" do
    assert {:ok, socket} = start_socket()
    assert :connected == TestSocket.wait_connect_status(socket)
    transport_pid = :sys.get_state(socket).transport_pid
    GenServer.stop(transport_pid)
    assert {:disconnected, {:transport_down, :normal}} == TestSocket.wait_connect_status(socket)

    # verify that we can reconnect and use the socket
    TestSocket.connect(socket)
    assert :connected == TestSocket.wait_connect_status(socket)
    assert {:ok, {"channel:1", _}} = TestSocket.join(socket, "channel:1")
  end

  test "can't interact when disconnected" do
    assert {:ok, socket} = start_socket()
    transport_pid = :sys.get_state(socket).transport_pid
    GenServer.stop(transport_pid)
    assert {:disconnected, _} = TestSocket.wait_connect_status(socket)
    assert {:error, :disconnected} = TestSocket.join(socket, "channel:1")
    assert {:error, :disconnected} = TestSocket.leave(socket, "channel:1")
    assert {:error, :disconnected} = TestSocket.push(socket, "channel:1", "some_event")
    assert {:error, :disconnected} = TestSocket.push_sync(socket, "channel:1", "some_event")
  end

  test "refused channel join" do
    assert {:ok, socket} = start_socket()
    assert :connected == TestSocket.wait_connect_status(socket)
    assert {:error, reason} = TestSocket.join(socket, "invalid_channel")
    assert {:server_rejected, "invalid_channel", %{"reason" => "unmatched topic"}} == reason
  end

  test "double join" do
    conn = join_channel()
    TestSocket.join(conn.socket, "channel:1")
    assert {:error, :already_joined} == TestSocket.join(conn.socket, "channel:1")
  end

  test "no push before join" do
    assert {:ok, socket} = start_socket()
    assert :connected == TestSocket.wait_connect_status(socket)
    assert {:error, :not_joined} == TestSocket.push(socket, "channel:1", "some_event")

    # Verify that we can still join after the invalid push
    assert {:ok, {"channel:1", _}} = TestSocket.join(socket, "channel:1")
    assert {:ok, _} = TestSocket.push(socket, "channel:1", "some_event")
  end

  test "server channel disconnects" do
    conn = join_channel()
    socket = conn.socket
    send(conn.server_channel, {:stop, :shutdown})
    assert_receive {^socket, :channel_closed, "channel:1", %{}}
  end

  test "server channel crashes" do
    conn = join_channel()
    socket = conn.socket

    ExUnit.CaptureLog.capture_log(fn ->
      send(conn.server_channel, {:crash, :some_reason})
      assert_receive {^socket, :channel_closed, "channel:1", %{}}
    end)
  end

  test "get status of joined channel" do
    conn = join_channel()
    assert TestSocket.joined?(conn.socket, "channel:1") == true
  end

  test "get status of not joined channel" do
    conn = join_channel()
    assert TestSocket.joined?(conn.socket, "channel:2") == false
  end

  defp join_channel do
    assert {:ok, socket} = start_socket()
    assert :connected == TestSocket.wait_connect_status(socket)
    assert {:ok, {"channel:1", %{}}} == TestSocket.join(socket, "channel:1")
    assert_receive {TestSite.Channel, {:join, "channel:1", _, server_channel}}

    %{socket: socket, server_channel: server_channel}
  end

  defp start_socket(url \\ url(), query_params \\ query_params(), connect \\ true) do
    TestSocket.start_link(
      Phoenix.Channels.GenSocketClient.Transport.WebSocketClient,
      url,
      query_params,
      connect
    )
  end

  defp url() do
    "#{Endpoint.url()}/test_socket/websocket"
    |> String.replace(~r(http://), "ws://")
    |> String.replace(~r(https://), "wss://")
  end

  defp query_params(), do: [{"shared_secret", "supersecret"}]

  defp url_updated() do
    "#{Endpoint.url()}/test_socket_updated/websocket"
    |> String.replace(~r(http://), "ws://")
    |> String.replace(~r(https://), "wss://")
  end

  defp query_params_updated(), do: [{"shared_secret", "supersecret_updated"}]
end


lib/gen_socket_client/transport.ex
defmodule Phoenix.Channels.GenSocketClient.Transport do
  @moduledoc """
  Transport contract used by `Phoenix.Channels.GenSocketClient`

  The implementation has following responsibilities:

  - Starts the transport process and links it to the caller. It is always assumed
    that the caller process is the socket process.
  - Pushes messages to the server when the `push/2` function is invoked.
  - Notifies the socket process about various events by invoking `notify_*`
    functions from the `Phoenix.Channels.GenSocketClient` module.
  """

  @type frame :: {:text | :binary, Phoenix.Channels.GenSocketClient.encoded_message()}
  @type transport_pid :: pid

  @doc "Invoked from the socket process to start the transport process."
  @callback start_link(url :: String.t(), Phoenix.Channels.GenSocketClient.transport_opts()) ::
              {:ok, transport_pid} | {:error, any}

  @doc "Invoked to push the frame."
  @callback push(transport_pid, frame) :: :ok
end


lib/gen_socket_client/serializer.ex
defmodule Phoenix.Channels.GenSocketClient.Serializer do
  @moduledoc """
  Describes the serializer interface used in `Phoenix.Channels.GenSocketClient`
  to encode/decode messages.
  """

  @doc "Invoked to decode the raw message."
  @callback decode_message(Phoenix.Channels.GenSocketClient.encoded_message(), Keyword.t()) ::
              Phoenix.Channels.GenSocketClient.message()

  @doc "Invoked to encode a socket message."
  @callback encode_message(Phoenix.Channels.GenSocketClient.message()) ::
              {:ok, Phoenix.Channels.GenSocketClient.Transport.frame()} | {:error, reason :: any}
end

defmodule Phoenix.Channels.GenSocketClient.Serializer.Json do
  @moduledoc "Json serializer for the socket client."
  @behaviour Phoenix.Channels.GenSocketClient.Serializer

  # -------------------------------------------------------------------
  # Phoenix.Channels.GenSocketClient.Serializer callbacks
  # -------------------------------------------------------------------

  @doc false
  def decode_message(encoded_message, _opts), do: Jason.decode!(encoded_message)

  @doc false
  def encode_message(message) do
    case Jason.encode(message) do
      {:ok, encoded} -> {:ok, {:text, encoded}}
      error -> error
    end
  end
end

defmodule Phoenix.Channels.GenSocketClient.Serializer.GzipJson do
  @moduledoc "Gzip+Json serializer for the socket client."
  @behaviour Phoenix.Channels.GenSocketClient.Serializer

  # -------------------------------------------------------------------
  # Phoenix.Channels.GenSocketClient.Serializer callbacks
  # -------------------------------------------------------------------

  @doc false
  def decode_message(encoded_message, _opts) do
    encoded_message
    |> :zlib.gunzip()
    |> Jason.decode!()
  end

  @doc false
  def encode_message(message) do
    case Jason.encode_to_iodata(message) do
      {:ok, encoded} -> {:ok, {:binary, :zlib.gzip(encoded)}}
      error -> error
    end
  end
end


lib/gen_socket_client/transport/web_socket_client.ex
defmodule Phoenix.Channels.GenSocketClient.Transport.WebSocketClient do
  @moduledoc """
  Websocket transport powered by the [websocket_client](https://github.com/sanmiguel/websocket_client)
  library.

  Supported transport options:

    - `keepalive` - Interval in which a ping message is sent to the server to keep the connection alive.
      By default, Phoenix server will timeout after 60 seconds of inactivity. By providing a keepalive value
      which is less than the server timeout, you can ensure that the connection remains open, even if  no
      messages are being passed between the client and the server. If you don't want to disable this mechanism,
      you can pass `nil`. If this option is not provided, the default value of 30 seconds is used.
    - `extra_headers` - List of headers to send in the handshake. It will be forwarded as is to the underlying library.
    - `ssl_verify` - It will be forwarded as is to the underlying library.
  """
  @behaviour Phoenix.Channels.GenSocketClient.Transport
  @behaviour :websocket_client
  @websocket_client_opts [:extra_headers, :ssl_verify]

  require Logger
  require Record
  alias Phoenix.Channels.GenSocketClient

  # -------------------------------------------------------------------
  # Phoenix.Channels.GenSocketClient.Transport callbacks
  # -------------------------------------------------------------------

  @doc false
  def start_link(url, transport_options) do
    {ws_opts, transport_options} = Keyword.split(transport_options, @websocket_client_opts)

    url
    |> to_charlist()
    |> :websocket_client.start_link(__MODULE__, [self(), transport_options], ws_opts)
  end

  @doc false
  def push(pid, frame) do
    send(pid, {:send_frame, frame})
    :ok
  end

  # -------------------------------------------------------------------
  # :websocket_client callbacks
  # -------------------------------------------------------------------

  @doc false
  def init([socket, transport_options]) do
    {:once,
     %{socket: socket, keepalive: Keyword.get(transport_options, :keepalive, :timer.seconds(30))}}
  end

  @doc false
  def onconnect(_req, state) do
    :ok = GenSocketClient.notify_connected(state.socket)

    case state.keepalive do
      nil -> {:ok, state}
      keepalive -> {:ok, state, keepalive}
    end
  end

  @doc false
  def websocket_handle({:pong, _message}, _req, state), do: {:ok, state}

  def websocket_handle({type, message}, _req, state) when type in [:text, :binary] do
    :ok = GenSocketClient.notify_message(state.socket, {type, message})
    {:ok, state}
  end

  def websocket_handle(other_msg, _req, state) do
    :ok = Logger.warn(fn -> "Unknown message #{inspect(other_msg)}" end)
    {:ok, state}
  end

  @doc false
  def websocket_info({:send_frame, frame}, _req, state), do: {:reply, frame, state}
  def websocket_info(_message, _req, state), do: {:ok, state}

  @doc false
  def ondisconnect(reason, state) do
    :ok = GenSocketClient.notify_disconnected(state.socket, reason)
    {:close, :normal, state}
  end

  @doc false
  def websocket_terminate(_reason, _req, _state), do: :ok
end


mix.exs
defmodule Phoenix.GenSocketClient.Mixfile do
  use Mix.Project

  @version "3.3.0"
  @github_url "https://github.com/alboratech/phoenix_gen_socket_client"
  @original_github_url "https://github.com/Aircloak/phoenix_gen_socket_client"

  def project do
    [
      app: :phoenix_gen_socket_client,
      version: @version,
      elixir: "~> 1.8",
      elixirc_paths: elixirc_paths(Mix.env()),
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      aliases: aliases(),
      deps: deps(),

      # Test
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        check: :test,
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],

      # Dialyzer
      dialyzer: dialyzer(),

      # Hex
      package: package(),
      description: "Socket client behaviour for phoenix channels.",

      # Docs
      docs: [
        source_url: @github_url,
        source_ref: "v#{@version}",
        main: "readme",
        extras: ["README.md"]
      ]
    ]
  end

  def application do
    [extra_applications: [:logger | extra_applications(Mix.env())]]
  end

  defp extra_applications(:prod), do: []
  defp extra_applications(_), do: [:websocket_client]

  defp deps do
    [
      {:websocket_client, "~> 1.2", optional: true},
      {:jason, "~> 1.2", optional: true},
      {:phoenix, "~> 1.5", only: :test},
      {:plug_cowboy, "~> 2.0", only: :test},
      {:credo, "~> 1.4", only: [:dev, :test], runtime: false},
      {:excoveralls, "~> 0.13", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 1.0", only: [:dev, :test], runtime: false},
      {:ex_doc, "~> 0.23", only: [:dev, :test], runtime: false}
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  defp aliases do
    [
      check: [
        "compile --warnings-as-errors",
        "format --check-formatted",
        "credo --strict",
        "coveralls.html",
        "dialyzer --format short"
      ]
    ]
  end

  defp package do
    [
      name: :phx_gen_socket_client,
      maintainers: ["Albora"],
      licenses: ["MIT"],
      links: %{
        "GitHub" => @github_url,
        "Original GitHub" => @original_github_url,
        "Docs" => "http://hexdocs.pm/phx_gen_socket_client"
      }
    ]
  end

  defp dialyzer do
    [
      plt_add_apps: [:mix, :ex_unit, :jason],
      plt_file: {:no_warn, "priv/plts/" <> plt_file_name()},
      flags: [
        :unmatched_returns,
        :error_handling,
        :race_conditions,
        :no_opaque,
        :unknown,
        :no_return
      ]
    ]
  end

  defp plt_file_name do
    "dialyzer-#{Mix.env()}-#{System.otp_release()}-#{System.version()}.plt"
  end
end
